### PR TITLE
Parse symmetry info and show in cell tab

### PR DESCRIPTION
**Pull request by unkcpz**
- [x] ~~Move the cell tab to front~~ Can be set when create widget.
- [x] parse symmetry info using `spglib`
- [x] ~~Using `ase.Atom` instead of `ase.Cell` in `StructureDataViewer` as active traitslet talk to other widget.~~
- [x] Missing unit symbol 'Angstrom'
- [x] ~~drop `ase.Cell` for structure viewer, use only `ase.Atoms` as trait.~~
- [x] When atom selected, automatically change to 'selection' tab.

### CODE REVIEW

**Comment by unkcpz** [comment]
fixes https://github.com/aiidalab/aiidalab-qe/issues/244

**Comment by yakutovicha** [comment]
@unkcpz I see there are still some non-completed tasks for this PR. Whenever this is ready please assign relevant people for the review.

**Comment by unkcpz** [comment]
There is still one thing probably good to have. When the structure is not chosen or uploaded the `StructureDataViewer` is hidden. If you all agree with this, since I am on this part of code, I can quickly add this feature to this PR.

**Comment by unkcpz** [comment]
The result of the widget after these change would be:
image

**Comment by csadorf** [comment]
@unkcpz Please re-request my review when this is ready, thank you!

**Comment by unkcpz** [comment]
@csadorf Yes, it is ready to review.

File: aiidalab_widgets_base/utils/__init__.py
Comment (csadorf) [inline aiidalab_widgets_base/utils/__init__.py:160]:
```suggestion
    """Convert ase Atoms instance to spglib cell."""
```

File: aiidalab_widgets_base/utils/__init__.py
Comment (csadorf) [inline aiidalab_widgets_base/utils/__init__.py:159]: Can you declare the return type here?

File: aiidalab_widgets_base/utils/__init__.py
Comment (csadorf) [inline aiidalab_widgets_base/utils/__init__.py:165]: Is this the definition of an spglib cell? Can we maybe add a link to the corresponding docs to the doc-string?

File: aiidalab_widgets_base/viewers.py
Comment (csadorf) [inline aiidalab_widgets_base/viewers.py:407]: Afaik the unit, unlike the name, is typically not capitalized.
```suggestion
                ipw.HTML("Length unit: angstrom"),
```

File: aiidalab_widgets_base/viewers.py
Comment (csadorf) [inline aiidalab_widgets_base/viewers.py:441]:
```suggestion
                                ipw.HTML("Symmetry information:"),
```

File: aiidalab_widgets_base/viewers.py
Comment (csadorf) [inline aiidalab_widgets_base/viewers.py:678]:
```suggestion
        # if atom is selected from nglview, shift to selection tab
```

File: aiidalab_widgets_base/viewers.py
Comment (csadorf) [inline aiidalab_widgets_base/viewers.py:171]: 1. Do we need to add a deprecation warning then?
2. What is the motivation for this?

File: aiidalab_widgets_base/viewers.py
Comment (csadorf) [inline aiidalab_widgets_base/viewers.py:161]: What are the allowed values here? Should we maybe provide an `Enum` of valid values?

File: aiidalab_widgets_base/viewers.py
Comment (csadorf) [inline aiidalab_widgets_base/viewers.py:180]: [One should not use mutable default values in Python.](https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/)

**Review by csadorf — CHANGES_REQUESTED** [review]

File: aiidalab_widgets_base/viewers.py
Comment (unkcpz) [reply aiidalab_widgets_base/viewers.py:171]: per 1: Maybe (only if other app use this widget and directly use `cell` traitslet, @yakutovicha if the apps from empa using this directly). 

per 2: The motivation to replace `cell` with `ase_structure` is to read the symmetry information from the traitlets . From my point of view, it makes more sense to store and use full structure.

File: aiidalab_widgets_base/viewers.py
Comment (unkcpz) [reply aiidalab_widgets_base/viewers.py:161]: These are four allowed values. I think `Enum` makes it more robust, will give it a try.

Comment (csadorf) [reply aiidalab_widgets_base/viewers.py:161]: "Cell" appears twice.

Comment (unkcpz) [reply aiidalab_widgets_base/viewers.py]: As we discussed in the meeting, not necessary change to `Enum` the scope of these options are all only used in this function for mapping.

File: aiidalab_widgets_base/utils/__init__.py
Comment (unkcpz) [reply aiidalab_widgets_base/utils/__init__.py]: Changed.

File: aiidalab_widgets_base/utils/__init__.py
Comment (unkcpz) [reply aiidalab_widgets_base/utils/__init__.py:169]: Link https://spglib.github.io/spglib/python-spglib.html#crystal-structure-cell added to docstring.

File: aiidalab_widgets_base/viewers.py
Comment (csadorf) [reply aiidalab_widgets_base/viewers.py:171]: I'll wait for @yakutovicha to comment on this.

Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py:171]: The `structure` trait of the `StructureDataViewer` child class serves the purpose of `ase_structure`. You can extract all the information from it.

> per 2: The motivation to replace `cell` with `ase_structure` is to read the symmetry information from the traitlets . From my point of view, it makes more sense to store and use full structure.

We actually thought of `_StructureDataBaseViewer` as a base class that could be used also for the trajectory viewers. I would, therefore, be very careful in changing this part.

I would at least do this particular change in another PR so that we could carefully consider all the pros and cons.

Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py:171]: Just to mention that @danielhollas is currently preparing things for the trajectory viewer, so we might want to revisit the design of `_StructureDataBaseViewer` class.

I can create a separate issue to collect the ideas. What do you think @unkcpz and @danielhollas?

File: aiidalab_widgets_base/viewers.py
Comment (danielhollas) [inline aiidalab_widgets_base/viewers.py:342]: @unkcpz  In this logic, please take into account non-periodic systems by checking the pbc attribute of the ase object.

File: aiidalab_widgets_base/viewers.py
Comment (unkcpz) [reply aiidalab_widgets_base/viewers.py:171]: Thanks @yakutovicha please open an issue for trajectory. 
I will revert this part in the pr. Basically I need the ase.Atom only for parse the symmetry info.

Comment (unkcpz) [reply aiidalab_widgets_base/viewers.py:173]: Look at the implementation of this PR again, if I revert back, there is no essential changes left.
I think if we call this class a 'StructureData' viewer, the positions info should be included with is not in the original implementation, which makes not much sense to me. 
@danielhollas Could you have a look at this implementation to see if this will conflict with your trajectory feature, if not, it is much easier to keep the base traitslet type change here. 
Therefore, I'll remove the `cell` traitslet if it is fine for you.

File: aiidalab_widgets_base/viewers.py
Comment (unkcpz) [reply aiidalab_widgets_base/viewers.py:344]: Good point! @danielhollas 

I add a check and only parse and attach symmetry when pbc shows it is 3-D

Comment (danielhollas) [reply aiidalab_widgets_base/viewers.py:344]: @unkcpz it might also be nice to automatically hide the Cell tab in that case. Not sure how much work that would be so perhaps in another PR?

File: aiidalab_widgets_base/viewers.py
Comment (danielhollas) [inline aiidalab_widgets_base/viewers.py:799]: I don't think this works in general case because the `structure` trait can be AiiDA node (StructureData or CifData). You would need to move this to the `displayed_structure` observer, but as mentioned in the other comment, I think it is easier to just move the `displayed_structure` trait to the parent class.

**Review by danielhollas — CHANGES_REQUESTED** [review]
@unkcpz I left a couple comments, let me know if my suggestions make sense.

File: aiidalab_widgets_base/viewers.py
Comment (unkcpz) [reply aiidalab_widgets_base/viewers.py:344]: Probably can update the tab widget but then need to bring it back when 3D structure is selected again, seems a bit tedious to make it work correctly, I'll take a look. 
Or we can make the widget support  the 2D and 1D with big cell set like we have for [optimade-client]?(https://optimadeclient.materialscloud.io/). From where if you search structure in MC2D database. Not sure I remember how we do that, I'll check it.

**Comment by unkcpz** [comment]
Thanks for the review! @danielhollas 
I think I am a bit confused with the design here. Can we have a short meeting to discuss it (we can meet next aiidalab meeting or quickly set a meeting if you are available)? 
The catch here is the `_StructureDataBaseViewer` has no positions info but I need to put the symmetry info in the `cell` tab which needs to use cell + position info. We need first to clarify the variable names used for our class since 'cell' sometimes only refers to lattice but in some libraries refers to the whole structure.

File: aiidalab_widgets_base/viewers.py
Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py:173]: > Look at the implementation of this PR again, if I revert back, there is no essential changes left.

Not really, I guess the purpose of this PR is to have the symmetry information displayed. This can stay.

>the positions info should be included with is not in the original implementation

sorry, I don't understand this, could you maybe rephrase?


> Just a correction to this, the structure trait can be either Atoms or AiiDA node, whereas `displayed_structure` is guaranteed to be the ASE Atoms object. So that's why I think it would make sense to move here.

No, it **_accepts_** both Atoms and AiiDA node, but then they get immediately converted into an ASE object:

https://github.com/aiidalab/aiidalab-widgets-base/blob/cabc5df932f41fcaf9907ac1ff2f0216ba20a700/aiidalab_widgets_base/viewers.py#L737-L738

> Regardless, I am fine with removing the cell traitlet. Since this is inside the private class, other things should not be using it anyway no?

It will also stay as a public member of the derived classes, we still need to be careful about removing it.

[P.S.]
Overall, I would like to stress that generally, I agree - we might change the classes to better satisfy our needs. We might do it in a separate PR (as I was suggesting) or just bite the bullet and do everything here. 

I try to be practical, that is why I suggest merging things that are clear and addressing the rest in a separate PR. If you want to do everything here - that is also fine with me. But keep in mind that it will cause some delays and long difficult-to-follow discussions.

**Comment by yakutovicha** [comment]
> Thanks for the review! @danielhollas I think I am a bit confused with the design here. Can we have a short meeting to discuss it (we can meet next aiidalab meeting or quickly set a meeting if you are available)? The catch here is the `_StructureDataBaseViewer` has no positions info but I need to put the symmetry info in the `cell` tab which needs to use cell + position info. We need first to clarify the variable names used for our class since 'cell' sometimes only refers to lattice but in some libraries refers to the whole structure.

I agree, we need to meet to clear things out.

Just as a brief mention: the idea to keep the `structure` trait outside of `_StructureDataBaseViewer` is coming from the generality principle. We wanted to reuse the same base class for the trajectory, where the trait would probably be called `trajectory`.

**Comment by danielhollas** [comment]
Sure, let's have a quick meeting. The next AiiDAlab meeting is after the Psi-k conference and I cannot attend unfortunately since I am at another conference in Rennes. Could we do a short meeting today or tomorrow afternoon?

@yakutovicha to clarify my suggestion, I would move the `displayed_structure` to the base class. That way there is a clearer separation of concern, since the base class is then in charge of what is displayed. The structure (or trajectory) traits would then stay in the concrete viewers. StructureDataViewer would then simply pass the structure info from structure trait to displayed_structure trait (converting to ase object from AiiDA node object), while the TrajectoryDataViewer would be passing the individual structure based on some selection widget.

**Comment by yakutovicha** [comment]
> Could we do a short meeting today or tomorrow afternoon?

I am available both today and tomorrow. @unkcpz what about you?

**Comment by unkcpz** [comment]
I have time today 3pm-3:30pm and tomorrow after 4pm. If you can find time, let me know. Thanks!

**Comment by danielhollas** [comment]
@unkcpz @yakutovicha I am available now but tomorrow works as well.

**Comment by yakutovicha** [comment]
> @unkcpz @yakutovicha I am available now but tomorrow works as well.

@danielhollas I've just sent you the link via email.

File: aiidalab_widgets_base/viewers.py
Comment (danielhollas) [reply aiidalab_widgets_base/viewers.py:344]:
```suggestion
        if self.ase_structure and any(self.ase_structure.pbc):
```

**Comment by unkcpz** [comment]
Hi @yakutovicha @danielhollas, I adapt the changes without touching the `cell` trait. I hope this is what we agreed on.

**Review by danielhollas — DISMISSED** [review]
@unkcpz thanks, the code looks good to me. I have some suggestions / questions, but those could be done in subsequent PR if needed.

I did not actually test this, I am leaving that to somebody with more solid state background.

File: aiidalab_widgets_base/viewers.py
Comment (unkcpz) [reply aiidalab_widgets_base/viewers.py:344]: Thanks for mentioning this. Actually, I made some effort to implement that but failed. The reason is that the code either hides the tab or adds a message inside the cell tab makes more sense to put in `_cell_tab` method, but in there the `self.structure` trait is not being set yet. 
Adding that too this PR will bring more coupling. But you are right, if it is okay with you, I'll open an issue for this and leave it to the future after we have decoupled the base structure data view.

File: aiidalab_widgets_base/viewers.py
Comment (unkcpz) [reply aiidalab_widgets_base/viewers.py:343]: There are two reasons that make this not easy to show. First, if it is a 2D structure, the spglib does not work well with fining the symmetry for 2D structure which has an [different set of space group](https://www.iucr.org/education/pamphlets/21/supplement). Secondly, I have no idea what to show for the cell information, do we still show a, b, c with c set to a large value as mimicking the vacuum layer, or don't show the c and beta in the tab.

**Review by unkcpz — COMMENTED** [review]
Thanks! @danielhollas

File: aiidalab_widgets_base/viewers.py
Comment (unkcpz) [reply aiidalab_widgets_base/viewers.py:344]: Test

Comment (yakutovicha) [reply aiidalab_widgets_base/viewers.py:344]: @unkcpz did you open the issue? If yes, please mention it here for the record

Comment (unkcpz) [reply aiidalab_widgets_base/viewers.py:344]: I just open the issue https://github.com/aiidalab/aiidalab-widgets-base/issues/351 for this.

**Review by yakutovicha — DISMISSED** [review]
All good also from my side, thanks @unkcpz!

**Comment by unkcpz** [comment]
Okay, a bit late than the PR of @danielhollas (3 min 🥺), introduce a conflict here. I resolve the conflict and rebase it and should be fine now. Nothing is charged so I'll merge it without your approval.

**Pull request merged**
